src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php

```diff
@@ -300,6 +291,457 @@ public static function analyze(
+        $parent_nodes = [];
+
+        if ($statements_analyzer->data_flow_graph instanceof VariableUseGraph) {
+            $parent_nodes = $stmt_type->parent_nodes;
+        }
+
+        while ($atomic_types) {
+            $atomic_type = array_pop($atomic_types);
+
+            if ($atomic_type instanceof TInt) {
+                $valid_ints[] = $atomic_type;
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TFloat) {
+                if ($atomic_type instanceof TLiteralFloat) {
+                    $valid_ints[] = new TLiteralInt((int) $atomic_type->value);
+                } else {
+                    $castable_types[] = new TInt();
+                }
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TString) {
+                if ($atomic_type instanceof TLiteralString) {
+                    $valid_ints[] = new TLiteralInt((int) $atomic_type->value);
+                } elseif ($atomic_type instanceof TNumericString) {
+                    $castable_types[] = new TInt();
+                } else {
+                    // any normal string is technically $valid_int[] = new TLiteralInt(0);
+                    // however we cannot be certain that it's not inferred, therefore less strict
+                    $castable_types[] = new TInt();
+                }
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TNull || $atomic_type instanceof TFalse) {
+                $valid_ints[] = new TLiteralInt(0);
+                continue;
+            }
+
+            if ($atomic_type instanceof TTrue) {
+                $valid_ints[] = new TLiteralInt(1);
+                continue;
+            }
+
+            if ($atomic_type instanceof TBool) {
+                $valid_ints[] = new TIntRange(0, 1);
+                continue;
+            }
+
+            // could be invalid, but allow it, as it is allowed for TString below too
+            if ($atomic_type instanceof TMixed
+                || $atomic_type instanceof TClosedResource
+                || $atomic_type instanceof TResource
+                || $atomic_type instanceof Scalar
+            ) {
+                $castable_types[] = new TInt();
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TNamedObject
+                || $atomic_type instanceof TObjectWithProperties
+            ) {
+                $intersection_types = [$atomic_type];
+
+                if ($atomic_type->extra_types) {
+                    $intersection_types = array_merge($intersection_types, $atomic_type->extra_types);
+                }
+
+                foreach ($intersection_types as $intersection_type) {
+                    if ($intersection_type instanceof TNamedObject) {
+                        $intersection_method_id = new MethodIdentifier(
+                            $intersection_type->value,
+                            '__tostring'
+                        );
```

Comment:
Of course `DOMDocument` causes issues... Are you sure it uses `__toString` though? I sort of figured `DOMDocument` had its own weird cast handling? Maybe we just have to add a special case where it's a `TNamedObject` with `strtolower($...->value) === "domdocument"` (and maybe `"domnode"`, or maybe we should check if it extends `"domnode"`)?

@orklah do you have any idea how `DOM` casting works? I've never looked into it.

Reply:
I think the best way forward at this point is to just pretend all casts from objects extending `SimpleXMLElement`, `DOMNode`, `GMP`, or `Decimal\Decimal` are valid. We can try to get more specific later if it's possible.

Reply:
Try `$statements_analyzer->getCodebase()->classExtends(...)`, should also make the check a lot easier compared to what you have now. Also, maybe call it `$pseudo_castable_classes` since it's not really related to `__toString`, and (just to be _really_ nitpicky) make it a private const instead of a static var since it shouldn't change.